### PR TITLE
Add tests for snapshotting.

### LINKED ISSUE #175: Write tests for snapshotting in workflow rewrite
AC: Unit tests for recursive snapshotting.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
Closes #175.

### CODE REVIEW

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowContext.kt [hunk 1]
```diff
@@ -175,6 +175,20 @@ fun <StateT : Any, OutputT : Any, ChildStateT : Any, ChildOutputT : Any, ChildRe
     ): ChildRenderingT = compose(child, Unit, key, handler)
 // @formatter:on
 
+/**
+ * Convenience alias of [WorkflowContext.compose] for workflows that don't take input or emit output.
```
Comment (rjrjr) [inline kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowContext.kt]: Didn't you already put this change up in another PR?

File: kotlin/workflow-host/src/main/java/com/squareup/workflow/internal/LifetimeTracker.kt [hunk 2]
```diff
@@ -61,9 +61,18 @@ internal class LifetimeTracker<FactoryT, KeyT : Any, DisposableT>(
    * [Starts][start] [factories] with [keys][getKey] not already present in this manager,
    * and [disposes+removes][dispose] disposables for those that aren't present in [factories].
    */
-  fun track(factories: Iterable<FactoryT>) {
+  fun track(factories: List<FactoryT>) {
     val latestKeys = factories.associateBy(getKey)
 
+    // If these sizes don't match, there was at least one duplicate key.
+    require(factories.size == latestKeys.size) {
+      val duplicates = factories.groupBy(getKey)
```
Comment (rjrjr) [inline kotlin/workflow-host/src/main/java/com/squareup/workflow/internal/LifetimeTracker.kt]: Have I mentioned today how much I love this language?

File: kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt [hunk 3]
```diff
@@ -308,4 +314,209 @@ class WorkflowNodeTest {
...
+    )
+    assertEquals("initial input|child input", restoredNode.compose(parentWorkflow, "foo"))
+    assertEquals("child input", restoredChildState)
+    assertEquals("initial input", restoredParentState)
+  }
+
+  @Test fun snapshot_counts() {
+    var snapshotCalls = 0
+    var restoreCalls = 0
+    // Track the number of times the snapshot is actually serialized, not snapshotState is called.
```
Comment (rjrjr) [inline kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt]: Nice.

File: kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowContext.kt (hunk 1 above)
Comment (zach-klippenstein) [reply kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowContext.kt]: There are a bunch, I added one more.

File: kotlin/workflow-host/src/main/java/com/squareup/workflow/internal/RealWorkflowContext.kt [hunk 4]
```diff
@@ -81,7 +81,7 @@ internal class RealWorkflowContext<StateT : Any, OutputT : Any>(
         key: String,
         handler: (ChildOutputT) -> WorkflowAction<StateT, OutputT>
       ): ChildRenderingT {
-  // @formatter:on
+    // @formatter:on
```
Comment (kirillzh) [inline kotlin/workflow-host/src/main/java/com/squareup/workflow/internal/RealWorkflowContext.kt]: til

File: kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/LifetimeTrackerTest.kt [hunk 5]
```diff
@@ -90,4 +91,27 @@ class LifetimeTrackerTest {
...
+    }
+    assertTrue("3×dup" in error.message!!)
+  }
+
+  @Test fun `track throws on multiple sets of duplicate keys`() {
+    val error = assertFailsWith<IllegalArgumentException> {
+      tracker.track(listOf("dup1", "dup2", "dup1", "dup2"))
+    }
+    assertTrue("2×dup1" in error.message!!)
+    assertTrue("2×dup2" in error.message!!)
```
Comment (kirillzh) [inline kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/LifetimeTrackerTest.kt]: nit: `assertNotNull` might help you avoid bangs (since it has nullability inside it) and give you a better error message in case `error.message` is `null`.
```kotlin
assertNotNull(error.message)
assertTrue("2×dup1" in error.message)
assertTrue("2×dup2" in error.message)
```

File: kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt [hunk 6]
```diff
@@ -308,4 +314,209 @@ class WorkflowNodeTest {
...
+      }
+
+      override fun snapshotState(state: String): Snapshot = Snapshot.write {
+        it.writeUtf8WithLength("child state:$state")
+      }
+
+      override fun restoreState(snapshot: Snapshot): String = snapshot.bytes.parse {
+        it.readUtf8WithLength()
+            .removePrefix("child state:")
+            .also { state -> restoredChildState = state }
```
Comment (kirillzh) [inline kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt]: `restoredChildState = it.eadUtf8WithLength()...`?

File: kotlin/workflow-host/src/main/java/com/squareup/workflow/internal/RealWorkflowContext.kt (hunk 4 above)
Comment (zach-klippenstein) [reply kotlin/workflow-host/src/main/java/com/squareup/workflow/internal/RealWorkflowContext.kt]: Support for this syntax is not on by default in IntelliJ unfortunately, but it'll break ktlint if you accidentally reformat without it.

File: kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt (hunk 6 above)
Comment (zach-klippenstein) [reply kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/WorkflowNodeTest.kt]: The primary purpose of this function is to return the restored state, I feel like `also` expresses that this is a side-effect.

File: kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/LifetimeTrackerTest.kt (hunk 5 above)
Comment (zach-klippenstein) [reply kotlin/workflow-host/src/test/java/com/squareup/workflow/internal/LifetimeTrackerTest.kt]: image

File: kotlin/workflow-testing/src/test/java/com/squareup/workflow/TreeWorkflow.kt [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    input: String,
+    state: String,
+    context: WorkflowContext<String, Nothing>
+  ): Rendering {
+    val childRenderings = children
+        .mapIndexed { index, child ->
+          val childRendering = context.compose(child, "$input[$index]", child.name)
+          Pair(child.name, childRendering)
+        }
+        .toMap()
```
Comment (kirillzh) [inline kotlin/workflow-testing/src/test/java/com/squareup/workflow/TreeWorkflow.kt]:
```kotlin
children
    .withIndex()
    .associate { (index, child) ->
      val childRendering = context.compose(child, "$input[$index]", child.name)
      child.name to childRendering
    }
```
😛

**Comment by zach-klippenstein** [comment]
This PR also fixed a testing bug, so should make it in 0.9.0.
